feat(@embark/test-runner): introduce artifacts.require API

PascalPrecht opened the pull request:
This commit adds a convenience API `artifacts.require(name)` that aims to make
requiring artifacts a little bit more straight forward.

Usage:

```
const SimpleStorage = artifacts.require('SimpleStorage');
const EmbarkJS = artifacts.require('EmbarkJS');
```

### What did you refactor, implement, or fix?

<!-- Fill in the relevant information below to help us evaluate your proposed changes. -->

#### How did you do it?

<!-- Provide a summary of the improvement/s you are submitting. -->

#### Is there anything that needs special attention (breaking changes, etc)?

<!-- Explain any special considerations. -->

### Questions

<!-- If relevant, write a list of questions that you would like to discuss related to your changes. -->

### Review

<!-- Use @mentions for quick questions, specific feedback, and progress updates. -->

### Cool Spaceship Picture

<!-- Star Trek/Wars, n/BSG, Voltron, Transformers... all welcome, but not all equally cool ;-) ->

michaelsbradleyjr commented on packages/plugins/mocha-tests/src/lib/index.js:
```diff
@@ -222,6 +222,15 @@ class MochaTestRunner {
+                  if (name === 'EmbarkJS') {
+                    return EmbarkJS;
+                  }
+                  return require(`Embark/contracts/${name}`);
```
`Embark` is a special alias that's only setup in the context of a dapp when it's using the legacy webpack pipeline:
* https://github.com/embarklabs/embark/blob/3df75f12354e12ec09079bada287d64908235492/packages/plugins/basic-pipeline/src/webpack.config.js#L143-L145
* https://github.com/embarklabs/embark/blob/3df75f12354e12ec09079bada287d64908235492/packages/plugins/basic-pipeline/src/index.js#L52-L53

I'm confused re: how this is supposed to work because isn't mocha-tests something totally independent from the legacy pipeline?

jrainville replied:
Yes, but the tests have their own require override https://github.com/embarklabs/embark/blob/master/packages/plugins/mocha-tests/src/lib/index.js#L188-L211

michaelsbradleyjr replied:
Thanks @jrainville! I had forgotten about that, will unmark as "changes requested".

In the bigger scheme of things, I think this kind of patching is problematic, though it works for now and that's great. It's "magic" that goes against the idea of Embark being agnostic with respect to other tools, whether they be front-end build tools or test runners, etc. I think part of the future re-think of EmbarkJS needs to carefully consider the relationship between artifact/scaffolding generation and the range of ways users may need/want to import the generated modules: `require`, `import`, `import()`.

PascalPrecht replied:
> It's "magic" that goes against the idea of Embark being agnostic with respect to other tools, whether they be front-end build tools or test runners, etc

@michaelsbradleyjr I fully agree with what you say. Have the same concerns and raised them a couple of times in the past. However it seems that for now the main focus it to ensure interoperability/easy migration with other tools.

michaelsbradleyjr replied:
A few more thoughts on this, and maybe @jrainville can weigh in with respect to the [relevant code in mocha-tests](https://github.com/embarklabs/embark/blob/master/packages/plugins/mocha-tests/src/lib/index.js#L188-L211).

The use of `"Embark/contracts/"` in `mocha-tests` seems to have been for parity with the aliases setup in the legacy pipeline.

While removing the `Module.prototype.require` hack would be a breaking change — since dapps would no longer be able to manually `require('Embark/contracts/[contract]')` — maybe we have the opportunity to create a deprecation path.

What if an `artifactsRequire` function were created in an outer scope (or maybe as a static method on class `MochaTestRunner`). Then `global.artifacts.require` could simply be a reference to `artifactsRequire`; and `Module.prototype.require` could have logic that checks for `Embark/contracts` and `Embark/EmbarkJS` paths and in those cases call to `artifactsRequire` but warn the user that `Embark/` is deprecated and in the future global `artifacts.require()` should be used instead.

jrainville replied:
It should be doable. I don't mind. Depends what @iurimatias has in mind.
iirc, in the 4.0 breaking changes, we had that we no longer supported magic requires, but then we added them back 🤷‍♂ (confirmed here: https://github.com/embarklabs/embark/issues/2219)

michaelsbradleyjr replied:
Just to be clear. What I'm suggesting above is that since the magic requires are available to users currently, we shouldn't break them in v5. But we can move toward isolating them to the legacy webpack pipeline, since they're not in any way needed for a global `artifacts.require` to work as desired.

PascalPrecht commented:
Updated the PR to make CI happy. @michaelsbradleyjr I've also added an inline comment to avoid confusion for the next person looking at this code.

Thanks for the review!

emizzle commented:
I agree that we should create a deprecation path for magic imports, however, I don't think that needs to be addressed in this PR as the scope of changes is larger (would affect dapps and tests, whereas this PR affects only tests) and more considerations needs to be taken in to account.

michaelsbradleyjr commented:
Here's a comparison between this PR's branch and a branch that implements what I proposed in earlier comments:

https://github.com/embarklabs/embark/compare/feat/artifact-require...feat/artifact-require-mb

In addition to changes in `packages/plugins/mocha-tests/src/lib/index.js`, the dapp templates and tests have been updated to use the `artifacts.require()` syntax.

If this PR were updated to include the changes I'm proposing and included in Embark v5.2.0, then when `embark test` is run in dapps that use the "magic prefix" requires, a deprecation warning will be printed. See the screenshot below.

![image](https://user-images.githubusercontent.com/194260/73973186-be906200-48e7-11ea-80db-0f1954dc0f04.png)

In Embark v6.0.0 (or a later major version) we would then completely remove the monkey-patching of `Module.prototype.require` in `packages/plugins/mocha-tests/src/lib/index.js`.

iurimatias commented:
IMO we should go with @michaelsbradleyjr changes

PascalPrecht commented:
@iurimatias rebased this with @michaelsbradleyjr changes
